apps/shared/avifjpeg.c

```diff
@@ -266,6 +277,568 @@ static const uint8_t * avifJPEGFindSubstr(const uint8_t * str, size_t strLength,
+    AVIF_CHECK(avifJPEGFindGainMapPropertyDoubles(descNode,
+                                                  "HDRCapacityMax",
+                                                  /*log2Encoded=*/AVIF_TRUE,
+                                                  &metadataDouble.hdrCapacityMax,
+                                                  /*numDoubles=*/1));
+    AVIF_CHECK(avifJPEGFindGainMapPropertyDoubles(descNode, "OffsetSDR", /*log2Encoded=*/AVIF_FALSE, metadataDouble.offsetSdr, /*numDoubles=*/3));
+    AVIF_CHECK(avifJPEGFindGainMapPropertyDoubles(descNode, "OffsetHDR", /*log2Encoded=*/AVIF_FALSE, metadataDouble.offsetHdr, /*numDoubles=*/3));
+    AVIF_CHECK(avifJPEGFindGainMapPropertyDoubles(descNode,
+                                                  "GainMapMin",
+                                                  /*log2Encoded=*/AVIF_TRUE,
+                                                  metadataDouble.gainMapMin,
+                                                  /*numDoubles=*/3));
+    AVIF_CHECK(avifJPEGFindGainMapPropertyDoubles(descNode,
+                                                  "GainMapMax",
+                                                  /*log2Encoded=*/AVIF_TRUE,
+                                                  metadataDouble.gainMapMax,
+                                                  /*numDoubles=*/3));
+    AVIF_CHECK(avifJPEGFindGainMapPropertyDoubles(descNode,
+                                                  "Gamma",
+                                                  /*log2Encoded=*/AVIF_FALSE,
+                                                  metadataDouble.gainMapGamma,
+                                                  /*numDoubles=*/3));
+    // Change gammma to 1.0/gamma (we are using a different convetion from the metadata XMP).
+    for (int i = 0; i < 3; ++i) {
+        if (metadataDouble.gainMapGamma[i] == 0) {
+            return AVIF_FALSE; // Invalid value, the spec says that gamma should be > 0
+        }
+        metadataDouble.gainMapGamma[i] = 1.0 / metadataDouble.gainMapGamma[i];
+    }
+    // Should have HDRCapacityMax > HDRCapacityMin.
+    if (metadataDouble.hdrCapacityMax <= (double)metadataDouble.hdrCapacityMin) {
+        return AVIF_FALSE;
+    }
+    // Should have GainMapMax >= GainMapMin.
+    for (int i = 0; i < 3; ++i) {
+        if (metadataDouble.gainMapMax[i] < (double)metadataDouble.gainMapMin[i]) {
+            return AVIF_FALSE;
+        }
+    }
+
+    AVIF_CHECK(avifGainMapMetadataDoubleToFractions(metadata, &metadataDouble));
+
+    return AVIF_TRUE;
+}
+
+// Parses gain map metadata from an XMP payload.
+// Returns AVIF_TRUE if the gain map metadata was successfully read.
+static avifBool avifJPEGParseGainMapXMP(const uint8_t * xmpData, size_t xmpSize, avifGainMapMetadata * metadata)
+{
+    xmlDoc * document = xmlReadMemory((const char *)xmpData, (int)xmpSize, NULL, NULL, LIBXML2_XML_PARSING_FLAGS);
```

Comment:
It returns NULL from what I can tell. I added a check for that for clarity but I think it already worked because the next functions all return NULL if the input is NULL.